File: android/app/src/main/java/com/albbiz/map/StoriesRepository.kt
// Bismillah Hir Rahman Nir Raheem
package com.albbiz.map.data

import android.net.Uri
import android.util.Log
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import com.google.firebase.storage.FirebaseStorage
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.tasks.await

class StoriesRepository {
    private val db = FirebaseFirestore.getInstance()
    private val storiesRef = db.collection("stories")
    private val storage = FirebaseStorage.getInstance()

    // ── GET ACTIVE STORIES (not expired) ─────────────────────────────
    fun getActiveStories(): Flow<List<Story>> = callbackFlow {
        val now = System.currentTimeMillis()
        val listener = storiesRef
            .whereGreaterThan("expiresAt", now)
            .orderBy("expiresAt", Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, error ->
                if (error != null) {
                    close(error)
                    return@addSnapshotListener
                }
                val stories = snapshot?.documents?.mapNotNull { doc ->
                    doc.data?.let { Story.fromMap(doc.id, it) }
                } ?: emptyList()
                trySend(stories)
            }
        awaitClose { listener.remove() }
    }.catch { e ->
        Log.e("AlbBizMap", "StoriesRepo: Error getting stories", e)
        emit(emptyList())
    }

    // ── ADD STORY ─────────────────────────────────────────────────────
    suspend fun addStory(story: Story, photoUris: List<Uri>): Result<String> {
        return try {
            val storyRef = storiesRef.document()
            val storyId = storyRef.id

            // Upload photos to Firebase Storage
            val photoUrls = photoUris.mapIndexed { index, uri ->
                val filename = "stories/$storyId/photo_$index.jpg"
                val ref = storage.reference.child(filename)
                ref.putFile(uri).await()
                ref.downloadUrl.await().toString()
            }

            val finalStory = story.copy(
                id = storyId,
                photos = photoUrls
            )

            storyRef.set(finalStory.toMap()).await()
            Result.success(storyId)
        } catch (e: Exception) {
            Log.e("AlbBizMap", "StoriesRepo: Error adding story", e)
            Result.failure(e)
        }
    }

    // ── MARK STORY AS VIEWED ──────────────────────────────────────────
    suspend fun markStoryViewed(storyId: String, userId: String): Result<Unit> {
        return try {
            storiesRef.document(storyId)
                .update("viewedBy", com.google.firebase.firestore.FieldValue.arrayUnion(userId))
                .await()
            Result.success(Unit)
        } catch (e: Exception) {
            Log.e("AlbBizMap", "StoriesRepo: Error marking story viewed", e)
            Result.failure(e)
        }
    }

    // ── DELETE STORY ──────────────────────────────────────────────────
    suspend fun deleteStory(storyId: String): Result<Unit> {
        return try {
            storiesRef.document(storyId).delete().await()
            Result.success(Unit)
        } catch (e: Exception) {
            Log.e("AlbBizMap", "StoriesRepo: Error deleting story", e)
            Result.failure(e)
        }
    }

    // ── DELETE EXPIRED STORIES ────────────────────────────────────────
    // Only ever attempts to delete the current user's own expired stories.
    // This used to query every expired story app-wide and try to delete
    // all of them, but Firestore's security rules correctly block a user
    // from deleting someone else's document — so any story that wasn't
    // yours failed with PERMISSION_DENIED. Worse, since a single failed
    // delete() throws mid-loop, one other person's story appearing before
    // yours in the results could silently stop your own from being
    // cleaned up too, depending on document order. Filtering the query to
    // just this user's stories means every document it returns is one
    // it's actually allowed to delete, so the loop never aborts partway.
    // Note: this is a compound query (userId == X AND expiresAt < now) —
    // if it ever fails with a "requires an index" error, Firestore's own
    // error message includes a direct link to create that index in the
    // Firebase Console.
    suspend fun deleteExpiredStories(userId: String): Result<Unit> {
        if (userId.isBlank()) return Result.success(Unit)
        return try {
            val now = System.currentTimeMillis()
            val expired = storiesRef
                .whereEqualTo("userId", userId)
                .whereLessThan("expiresAt", now)
                .get()
                .await()
            expired.documents.forEach { it.reference.delete().await() }
            Result.success(Unit)
        } catch (e: Exception) {
            Log.e("AlbBizMap", "StoriesRepo: Error deleting expired stories", e)
            Result.failure(e)
        }
    }
}

File: ios/AlbBizMap_IOS/AlbBizMap/Services/StoriesRepository.swift
// Bismillah Hir Rahman Nir Raheem
import Foundation
import Combine
import FirebaseFirestore
import FirebaseStorage

public class StoriesRepository: ObservableObject {
    public static let shared = StoriesRepository()

    @Published public var stories: [Story] = []

    private let db = Firestore.firestore()
    private lazy var storiesRef = db.collection("stories")
    private let storage = Storage.storage()

    private var storiesListener: ListenerRegistration?

    private init() {
        startListening()
    }

    // Mirrors Android's getActiveStories: only stories that haven't expired
    // yet, newest expiry first.
    private func startListening() {
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        storiesListener = storiesRef
            .whereField("expiresAt", isGreaterThan: now)
            .order(by: "expiresAt", descending: true)
            .addSnapshotListener { [weak self] snapshot, error in
                guard let self else { return }
                if let error {
                    print("StoriesRepository: error listening for stories: \(error)")
                    return
                }
                let stories: [Story] = snapshot?.documents.compactMap { doc in
                    Story.fromMap(id: doc.documentID, map: doc.data())
                } ?? []
                DispatchQueue.main.async {
                    self.stories = stories
                }
            }
    }

    // Mirrors Android's addStory(story, photoUris): uploads local photo data
    // to Storage first, then writes the story with those URLs.
    public func addStory(_ story: Story, photoData: [Data] = []) async -> Result<String, Error> {
        let docRef = storiesRef.document()
        do {
            var photoUrls: [String] = []
            for (index, data) in photoData.enumerated() {
                let filename = "stories/\(docRef.documentID)/photo_\(index).jpg"
                let ref = storage.reference().child(filename)
                let metadata = StorageMetadata()
                metadata.contentType = "image/jpeg"
                _ = try await ref.putDataAsync(data, metadata: metadata)
                let url = try await ref.downloadURL()
                photoUrls.append(url.absoluteString)
            }
            var finalStory = story
            finalStory.id = docRef.documentID
            finalStory.photos = photoUrls
            try await docRef.setData(finalStory.toMap().compactMapValues { $0 })
            return .success(docRef.documentID)
        } catch {
            return .failure(error)
        }
    }

    // addStory() above expects raw photo data it still needs to upload — for
    // the auto-generated "Just opened" story created right after a business
    // is registered, the photos are the business's own photos, already
    // uploaded to Storage as URLs at that point. Mirrors Android's
    // addStoryWithHostedPhotos, which skips the upload step entirely so the
    // story keeps the photos it was already given instead of them being
    // overwritten by an empty upload result.
    public func addStoryWithHostedPhotos(_ story: Story) async -> Result<String, Error> {
        let docRef = storiesRef.document()
        do {
            var finalStory = story
            finalStory.id = docRef.documentID
            try await docRef.setData(finalStory.toMap().compactMapValues { $0 })
            return .success(docRef.documentID)
        } catch {
            return .failure(error)
        }
    }

    public func markStoryViewed(storyId: String, userId: String) {
        Task {
            do {
                try await storiesRef.document(storyId).updateData([
                    "viewedBy": FieldValue.arrayUnion([userId])
                ])
            } catch {
                print("StoriesRepository: error marking story viewed: \(error)")
            }
        }
    }

    public func deleteStory(storyId: String) {
        Task {
            do {
                try await storiesRef.document(storyId).delete()
            } catch {
                print("StoriesRepository: error deleting story: \(error)")
            }
        }
    }

    // Only ever deletes the current user's own expired stories — matches
    // Android's scoping, since Firestore security rules block deleting
    // someone else's document and a compound query keeps every deletion
    // this loop attempts one it's actually allowed to make.
    public func deleteExpiredStories(userId: String) async {
        guard !userId.isEmpty else { return }
        do {
            let now = Int64(Date().timeIntervalSince1970 * 1000)
            let snapshot = try await storiesRef
                .whereField("userId", isEqualTo: userId)
                .whereField("expiresAt", isLessThan: now)
                .getDocuments()
            for doc in snapshot.documents {
                try await doc.reference.delete()
            }
        } catch {
            print("StoriesRepository: error deleting expired stories: \(error)")
        }
    }
}
